### PR TITLE
ENH: add code for marron wand simulator

### PR DESCRIPTION
**Pull request by sampan501**
N/A

Changes proposed in this pull request:

- Added Marron and Wand 1992 simulations with multivariate normal simulations
- Added truth for each simulation

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: sktree/datasets/hyppo.py [hunk 1]
```diff
@@ -47,14 +47,33 @@ def make_quadratic_classification(n_samples: int, n_features: int, noise=False,
     return x, v
 
 
+MARRON_WAND_SIMS = {
```
Comment (adam2392) [inline sktree/datasets/hyppo.py:53]: Can you add a docstring above describing what the list is comprised of and what the simulations are?

File: sktree/datasets/hyppo.py [hunk 2]
```diff
@@ -110,8 +133,9 @@ def make_trunk_classification(
     """
     rng = np.random.default_rng(seed=seed)
 
-    mu_1 = np.array([1 / np.sqrt(i) for i in range(1, n_dim + 1)])
-    mu_0 = m_factor * mu_1
+    mu_1 = m_factor * np.array([1 / np.sqrt(i) if i < 257 else 0 for i in range(1, n_dim + 1)])
```
Comment (adam2392) [inline sktree/datasets/hyppo.py]: Where did the 257 come from? Is that the hyperparameter to limit the number of informative dimensions?

In #200 I implemented `n_dim` vs `n_informative`, which I think is more general and self-documenting.

File: sktree/datasets/hyppo.py [hunk 3]
```diff
@@ -123,36 +147,280 @@ def make_trunk_classification(
...
         )
+    else:
+        raise ValueError(
+            f"Simulation must be trunk, trunk_overlap, trunk_mix, {MARRON_WAND_SIMS.keys()}"
+        )
+
     y = np.concatenate((np.zeros(n_samples // 2), np.ones(n_samples // 2)))
 
     if return_params:
         return X, y, [mu_0, mu_1], [cov, cov]
```
Comment (adam2392) [inline sktree/datasets/hyppo.py]: Technically, the mu and covariance are not easily returned anymore. I'm wondering if we should wrap them in an object then instead?

**Review by adam2392 — COMMENTED** [review]
Few comments that I think can result in runtime issues and hidden bugs even.

Can you also add a unit test? I think an easy thing to test is just that things run, expected error comes out for `simulation` and

**Comment by adam2392** [comment]
Lmk when this is ready for review.

Now `n_informative` controls the number of informative features. You can set it to default 256. cc: https://github.com/neurodata/scikit-tree/pull/203#discussion_r1466439669

File: sktree/datasets/hyppo.py (hunk 3 above)
Comment (sampan501) [reply sktree/datasets/hyppo.py]: I'm not sure, what do you think we need to return in `return_params`?

Comment (adam2392) [reply sktree/datasets/hyppo.py]: I guess we use `return_params` to get information about the data generating model as well as possibly parameters of the DGM.

Is it easy to wrap these as a subclass for `rv_continuous` class (https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.html#scipy.stats.rv_continuous)? Then we can just return the instantiated object. Though I'm not familiar with that API, and how we can recover the exact parametrization of a specific MW simulation.

**Comment by sampan501** [comment]
@adam2392 PR is ready except for formatting. Not sure where the error is

**Comment by adam2392** [comment]
You can run `make pre-commit` to run pre-commit linting pipelines, or setup pre-commit to automatically run before each commit.

I pushed [665dfdf](https://github.com/neurodata/scikit-tree/pull/203/commits/665dfdf5bb11e303007d26622aa8c61bddf519e9) which should fix the issues

File: sktree/datasets/hyppo.py (hunk 3 above)
Comment (adam2392) [reply sktree/datasets/hyppo.py]: @sampan501 if there's no way to address this, then should we just error out on `return_params=True` if simulation is not trunk?

I think we just want the DGM to be returned, so perhaps it would make sense to just return a callable such that
`returned_model.sample(n_samples)` can generate new samples of data. For `rng.multivariate_normal` or if we can get an API in the framework of `rv_continuous`, then we can do things like:

```
test = multivariate_normal(mean=2.5, cov=0.5)
print(test)
print(test.mean)
print(test.cov)
print(test.rvs(size=(10,)))
```

File: sktree/datasets/hyppo.py [hunk 4]
```diff
@@ -121,6 +153,7 @@ def make_trunk_classification(
 
     mu_1 = np.array([1 / np.sqrt(i) for i in range(1, n_informative + 1)])
     mu_0 = m_factor * mu_1
+    w = mu_1
```
Comment (adam2392) [inline sktree/datasets/hyppo.py]: What is `w`?

File: sktree/datasets/hyppo.py [hunk 5]
```diff
@@ -121,6 +153,7 @@ def make_trunk_classification(
 
     mu_1 = np.array([1 / np.sqrt(i) for i in range(1, n_informative + 1)])
     mu_0 = m_factor * mu_1
```
Comment (adam2392) [inline sktree/datasets/hyppo.py]: Mention in docstring that `m_factor` is only used for `trunk` simulation and `trunk_overlap` is equivalent to `m_factor = 1`

File: sktree/datasets/hyppo.py [hunk 6]
```diff
@@ -91,8 +116,15 @@ def make_trunk_classification(
...
     return_params : bool, optional
         Whether or not to return the distribution parameters of the classes normal distributions.
-    mix : float, optional
-        Whether or not to mix the Gaussians. Should be a value between 0 and 1.
+    simulation : str, optional
+        Which simulation to run. Must be: 'trunk', 'trunk_overlap', 'trunk_mix', or one of the
+        following Marron-Wand simulations: 'gaussian', 'skewed_unimodal', 'strongly_skewed',
+        'kurtotic_unimodal', 'outlier', 'bimodal', 'separated_bimodal', 'skewed_bimodal',
+        'trimodal', 'claw', 'double_claw', 'asymmetric_claw', 'asymmetric_double_claw',
+        'smooth_comb', 'discrete_comb'.
```
Comment (adam2392) [inline sktree/datasets/hyppo.py]: Can you describe here, or in the `Notes` section of the docstring how `mu_1, cov` is used in Marron-Wand simulations?

It seems that `cov` is passed to Marron and Wand

**Review by adam2392 — COMMENTED** [review]
Since we'll be using the simulations a lot, it would be great to document it well at this stage to prevent future confusion. I left a few comments on where I think it would be necessary.

File: sktree/datasets/hyppo.py (hunk 5 above)
Comment (sampan501) [reply sktree/datasets/hyppo.py]: https://github.com/neurodata/scikit-tree/pull/203/commits/b1a0dcaf3289e1209838236547203751a41f4db0

File: sktree/datasets/hyppo.py (hunk 4 above)
Comment (sampan501) [reply sktree/datasets/hyppo.py]: https://github.com/neurodata/scikit-tree/pull/203/commits/b1a0dcaf3289e1209838236547203751a41f4db0

File: sktree/datasets/hyppo.py (hunk 6 above)
Comment (sampan501) [reply sktree/datasets/hyppo.py]: https://github.com/neurodata/scikit-tree/pull/203/commits/b1a0dcaf3289e1209838236547203751a41f4db0

File: sktree/datasets/hyppo.py (hunk 3 above)
Comment (sampan501) [reply sktree/datasets/hyppo.py]: we could just return `X`, `Y`, `norm_params`, and `G` for the Marron and Wand Sims right?

Comment (sampan501) [reply sktree/datasets/hyppo.py]: The only thing additional is G, which are the observations from the mixed Gaussian

Comment (adam2392) [reply sktree/datasets/hyppo.py]: Yeah that sounds reasonable for now. Can you update the docstring accordingly?

Comment (sampan501) [reply sktree/datasets/hyppo.py]: https://github.com/neurodata/scikit-tree/pull/203/commits/6688725b70666c5b86b3d4c87f1cba316340eeca

File: sktree/datasets/hyppo.py [hunk 7]
```diff
@@ -135,49 +180,228 @@ def make_trunk_classification(
...
 
     if return_params:
-        return X, y, [mu_0, mu_1], [cov, cov]
+        returns = [X, y]
+        if simulation == "trunk":
+            returns += [[mu_0, mu_1], [cov, cov]]
+        elif simulation == "trunk-overlap":
+            returns += [[np.zeros(n_informative), np.zeros(n_informative)], [cov, cov]]
+        elif simulation == "trunk-overlap":
+            returns += [*list(zip(*norm_params)), X_mixture]
```
Comment (adam2392) [inline sktree/datasets/hyppo.py]: Aren't these the same as L271

File: sktree/datasets/hyppo.py [hunk 8]
```diff
@@ -135,49 +180,228 @@ def make_trunk_classification(
...
+        else:
+            returns += [*list(zip(*norm_params)), G, w]
```
Comment (adam2392) [inline sktree/datasets/hyppo.py:277]: What are `norm_params` exactly? Is it possible to describe where they are initialized what they are? Just because otw it's hard to parse.

**Comment by adam2392** [comment]
Few other nit picks and then LGTM

File: sktree/datasets/hyppo.py (hunk 8 above)
Comment (sampan501) [reply sktree/datasets/hyppo.py:277]: For trunk mix, `norm_params = [[mu, cov], [mu, cov]]` (L209) and the Marron Wand sims, it's the parameters used for each Gaussian in the mixture (L233).
I wrote it to match the returns that we get for the other simulations, which are `[[mu, mu], [cov, cov]]`

File: sktree/datasets/hyppo.py (hunk 7 above)
Comment (sampan501) [reply sktree/datasets/hyppo.py]: https://github.com/neurodata/scikit-tree/pull/203/commits/7f7034363a069a3207b5b9dc78cb540880e3faa5

File: sktree/datasets/hyppo.py [hunk 9]
```diff
@@ -135,49 +180,228 @@ def make_trunk_classification(
...
+        )
+    elif simulation in MARRON_WAND_SIMS.keys():
         mixture_idx = rng.choice(
-            [0, 1], n_samples // 2, replace=True, shuffle=True, p=[mix, 1 - mix]
+            len(MARRON_WAND_SIMS[simulation]),  # type: ignore
+            size=n_samples // 2,
+            replace=True,
+            p=MARRON_WAND_SIMS[simulation],
+        )
+        norm_params = MarronWandSims(n_dim=n_informative, cov=cov)(simulation)
```
Comment (adam2392) [inline sktree/datasets/hyppo.py:234]:
```suggestion
        # the parameters used for each Gaussian in the mixture for each Marron Wand simulation
        norm_params = MarronWandSims(n_dim=n_informative, cov=cov)(simulation)
```

**Review by adam2392 — APPROVED** [review]
I'll let you merge once CIs are green
